---
title: "Human eTACS - Overlap with human mTECs"
output: html_notebook
---

## Introduction
Jo has kindly provided human mTEC bulk RNAseq data - these will be useful to compare the TRA expression against for the LN DCs/eTACs.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(vegan))
suppressPackageStartupMessages(library(Rtsne))
suppressPackageStartupMessages(library(destiny))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(lsa))
suppressPackageStartupMessages(library(biomaRt))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(flashClust))
suppressPackageStartupMessages(library(goseq))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(VennDiagram))
suppressPackageStartupMessages(library(cowplot))
source('~/Dropbox/R_sessions/Clustering/CosineKNN.R')
source("~/Dropbox/R_sessions/GGMike/theme_mike.R")

etacs <- read.table("~/Dropbox/ETACS/human_eTAC-SFnorm.tsv",
                    h=T, sep="\t", stringsAsFactors=F)

etacs.meta <- read.table("~/Dropbox/ETACS/human_eTAC_pseudotime-metadata.tsv",
                         h=T, sep="\t", stringsAsFactors=F)
etacs.meta$CellType <- paste(etacs.meta$CellType, etacs.meta$MetaState, sep="_")

n.cells <- dim(etacs)[2] - 2
etacs.bin <- data.frame(t(apply(etacs[, 2:n.cells + 1],
                                1, FUN=function(B) binary_convert(B, 1))))
rownames(etacs.bin) <- etacs$gene_id
colnames(etacs.bin) <- colnames(etacs)[2:n.cells + 1]
etacs.bin$external_gene_name <- etacs$external_gene_name
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
mtec.bulk <- read.table("~/Dropbox/ETACS/human_etacs/diffexp.txt",
                        h=T, sep="\t", stringsAsFactors=F)
mtec.counts <- mtec.bulk[, 1:4]
rownames(mtec.counts) <- mtec.counts$ensembl_gene_id
mtec.counts <- mtec.counts[, -1]

nz <- rowSums(mtec.counts) > 3
mtec.nz <- mtec.counts[nz, ]
name.split <- strsplit(colnames(mtec.nz), fixed=T, split="_")

mtec.meta <- data.frame(cbind(colnames(mtec.nz), unlist(lapply(name.split, FUN=function(X) paste0(X[2]))),
                              c("mTEC", "mTEC", "mTEC")))
colnames(mtec.meta) <- c("Sample", "Replicate", "CellType")
rownames(mtec.meta) <- mtec.meta$Sample

dse <- DESeqDataSetFromMatrix(mtec.nz, colData=mtec.meta, design=~1)
dse <- estimateSizeFactors(dse)
mtec.norm <- data.frame(log2(counts(dse, normalized=TRUE) + 1))
mtec.norm$gene_id <- rownames(mtec.norm)
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
cDC.counts <- mtec.bulk[, c(1, 5:10)]
rownames(cDC.counts) <- cDC.counts$ensembl_gene_id
cDC.counts <- cDC.counts[, -1]

nz <- rowSums(cDC.counts) > 3
cDC.nz <- cDC.counts[nz, ]
name.split <- strsplit(colnames(cDC.nz), fixed=T, split="_")

cDC.meta <- data.frame(cbind(colnames(cDC.nz), unlist(lapply(name.split, FUN=function(X) paste0(X[2]))),
                              c("eTAC", "cDC", "eTAC", "cDC", "eTAC", "cDC")))
colnames(cDC.meta) <- c("Sample", "Replicate", "CellType")
rownames(cDC.meta) <- cDC.meta$Sample

dse <- DESeqDataSetFromMatrix(cDC.nz, colData=cDC.meta, design=~1)
dse <- estimateSizeFactors(dse)
cDC.norm <- data.frame(log2(counts(dse, normalized=TRUE) + 1))
cDC.norm$gene_id <- rownames(cDC.norm)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
source("~/Dropbox/R_sessions/eTAC/human_alpha_islets.R")
tra.df <- read.table("~/Dropbox/fantom5/hg19/Human-Tau_TRA_level3.tsv",
                     h=T, sep="\t", stringsAsFactors=F)

tra.quants <- quantile(tra.df$tau, probs=c(0.25, 0.75), na.rm=TRUE)
tra.df$GeneGroup <- "Misc"
tra.df$GeneGroup[tra.df$tau <= tra.quants[1]] <- "Constituitive"
tra.df$GeneGroup[tra.df$tau > tra.quants[2]] <- "Tissue-Restricted"

hum.merge <- Reduce(x=list("eTAC"=etacs, "Pancreas"=panc.alpha,
                           "mTEC"=mtec.norm, "cDC"=cDC.norm),
                    f=function(x, y) merge(x, y, by='gene_id', all=TRUE))
hum.merge[is.na(hum.merge)] <- 0

fantom.exprs <- merge(hum.merge, tra.df[, (dim(tra.df)[2]-5):(dim(tra.df)[2])],
                      by.x='gene_id', by.y='ensembl.gene')
fantom.cols <- colnames(tra.df[, (dim(tra.df)[2]-5):(dim(tra.df)[2])])[-2]
fant.melt <- melt(fantom.exprs, id.vars=c(fantom.cols, 'gene_id', 'external_gene_name'))
```




```{r, echo=FALSE, message=FALSE, warning=FALSE}
fantom.merge <- merge(fant.melt, etacs.meta[, c(1:3, 227:236)], by.x='variable', by.y='Sample', all=TRUE)
fantom.merge$CellType[grepl(fantom.merge$variable, pattern="alpha")] <- "human_AlphaIslet"
fantom.merge$CellType[grepl(fantom.merge$variable, pattern="mTEC")] <- "human_mTEC"
fantom.merge$CellType[grepl(fantom.merge$variable, pattern="cDC")] <- "cDC"
fantom.merge$CellType[grepl(fantom.merge$variable, pattern="_eTAC_")] <- "bulk_eTAC"
```

I have recreated the figure in a previous report comparing the proportion of genes expressed in each cell group based on the tissue-restricted pattern of expression.  The eTAC cell groups are from the meta-stable states in the pseudotime analysis.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=5.5, fig.width=7.5}
pge.human <- ggplot(fantom.merge[fantom.merge$value > 1  & !is.na(fantom.merge$CellType), ],
       aes(x=CellType, fill=GeneGroup)) +
  geom_bar(position='fill') + 
  theme_mike() + theme(axis.text.x=element_text(angle=90)) +
  scale_fill_Publication() +
  labs(x="Cell Type", y="Percentage of Expressed Genes")
ggsave(pge.human,
       filename="~/Dropbox/ETACS/plots.dir/human_eTAC-vs-human_mTEC-pGE.png",
       height=5.5, width=7.5, dpi=300)
pge.human
```

There is an increase in the mid-tissue restriction range of genes (`Misc`) in the yellow meta-stable state of LN cells.  These most likely represent a population of macrophages or monocytes, with high expression of CD14 and complement component genes.  Compared to the human mTECs we can see that there proportion of TRAs expressed is very low, and no different from what might be expected in peripheral tissues generally.

What about the actual TRAs expressed in mTEC vs LN DC subsets?  What is the degree of overlap, are they all already expressed in the thymus?

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=6, fig.width=9.5}
grid.newpage()
mtec.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("human_mTEC") & fantom.merge$GeneGroup == "Tissue-Restricted"])
yellow_etac.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_yellow") & fantom.merge$GeneGroup == "Tissue-Restricted"])
blue_etac.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_blue") & fantom.merge$GeneGroup == "Tissue-Restricted"])
brown_etac.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_brown") & fantom.merge$GeneGroup == "Tissue-Restricted"])
turq_etac.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_turquoise") & fantom.merge$GeneGroup == "Tissue-Restricted"])

tway.venn <- draw.quintuple.venn(area1=length(mtec.tras),
                                area2=length(yellow_etac.tras),
                                area3=length(brown_etac.tras),
                                area4=length(blue_etac.tras),
                                area5=length(turq_etac.tras),
                                n12=length(intersect(mtec.tras, yellow_etac.tras)),
                                n13=length(intersect(mtec.tras, brown_etac.tras)),
                                n14=length(intersect(mtec.tras, blue_etac.tras)),
                                n15=length(intersect(mtec.tras, turq_etac.tras)),
                                n23=length(intersect(yellow_etac.tras, brown_etac.tras)),
                                n24=length(intersect(yellow_etac.tras, blue_etac.tras)),
                                n25=length(intersect(yellow_etac.tras, turq_etac.tras)),
                                n34=length(intersect(brown_etac.tras, blue_etac.tras)),
                                n35=length(intersect(brown_etac.tras, turq_etac.tras)),
                                n45=length(intersect(blue_etac.tras, turq_etac.tras)),
                                n123=length(intersect(intersect(mtec.tras, yellow_etac.tras), brown_etac.tras)),
                                n124=length(intersect(intersect(mtec.tras, yellow_etac.tras), blue_etac.tras)),
                                n125=length(intersect(intersect(mtec.tras, yellow_etac.tras), turq_etac.tras)),
                                n134=length(intersect(intersect(mtec.tras, brown_etac.tras), blue_etac.tras)),
                                n135=length(intersect(intersect(mtec.tras, brown_etac.tras), turq_etac.tras)),
                                n145=length(intersect(intersect(mtec.tras, blue_etac.tras), turq_etac.tras)),
                                n234=length(intersect(intersect(yellow_etac.tras, brown_etac.tras), blue_etac.tras)),
                                n235=length(intersect(intersect(yellow_etac.tras, brown_etac.tras), turq_etac.tras)),
                                n245=length(intersect(intersect(yellow_etac.tras, blue_etac.tras), turq_etac.tras)),
                                n345=length(intersect(intersect(brown_etac.tras, blue_etac.tras), turq_etac.tras)),
                                n1234=length(intersect(intersect(intersect(mtec.tras, yellow_etac.tras), brown_etac.tras), blue_etac.tras)),
                                n1235=length(intersect(intersect(intersect(mtec.tras, yellow_etac.tras), brown_etac.tras), turq_etac.tras)),
                                n1245=length(intersect(intersect(intersect(mtec.tras, yellow_etac.tras), blue_etac.tras), turq_etac.tras)),
                                n1345=length(intersect(intersect(intersect(mtec.tras, brown_etac.tras), blue_etac.tras), turq_etac.tras)),
                                n2345=length(intersect(intersect(intersect(yellow_etac.tras, brown_etac.tras), blue_etac.tras), turq_etac.tras)),
                                n12345=length(intersect(intersect(intersect(intersect(yellow_etac.tras, brown_etac.tras), blue_etac.tras), turq_etac.tras), mtec.tras)),
                                category=c("mTEC", "Yellow\neTAC", "Brown\neTAC", "Blue\neTAC", "Turquoise\neTAC"),
                                lwd=c(1, 1, 1, 1, 1), col=rep("black", 5),
                                fill=c("#386cb0", "#fdb462", "#7fc97f","#ef3b2c","#662506"), alpha=rep(0.4, 5),
                                cat.pos=c(0, 320, 270, 120, 30))
# png("~/Dropbox/ETACS/plots.dir/eTAC_vs_mTEC-TRAs_Venn.png",
#     height=6, width=9.5, res=300, units="in")
grid.draw(tway.venn)
# dev.off()
```

Across all `tissue-restricted` genes, we can see that there are very few specific to each group, and that are not detected in the bulk mTEC RNAseq data.  The group that we might expect to see more TRAs is the `brown` eTAC group that express _AIRE_; this is the group which show the lowest number of specific/unique TRAs, suggesting the presence of _AIRE_ in these cells is not indicative of promiscuous gene expression.

This comparison is slightly biased towards the mTECs as the sensitivity for single cell RNAseq is much lower than for bulk RNAseq.  A properly valid comparison would be to look at the bulk RNAseq of these populations...

### Comparing the three bulk populations with TRAs only

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=8.5, fig.width=8.5}
grid.newpage()
cdc.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("cDC") & fantom.merge$GeneGroup == "Tissue-Restricted"])
bulk_etac.tras <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("bulk_eTAC") & fantom.merge$GeneGroup == "Tissue-Restricted"])

thway.venn <- draw.triple.venn(area1=length(mtec.tras),
                              area2=length(cdc.tras),
                              area3=length(bulk_etac.tras),
                              n12=length(intersect(mtec.tras, cdc.tras)),
                              n23=length(intersect(cdc.tras, bulk_etac.tras)),
                              n13=length(intersect(mtec.tras, bulk_etac.tras)),
                              n123=length(intersect(intersect(mtec.tras, cdc.tras), bulk_etac.tras)),
                                category=c("mTEC", "cDC", "bulk\neTAC"),
                                lwd=c(1, 1, 1), col=rep("black", 3),
                                fill=c("#386cb0", "#A50149", "#4E0025"), alpha=rep(0.4, 3),
                                cat.pos=c(330, 30, 180), cat.just=list("mTEC"=c(0, 0),
                                                                    "cDC"=c(0, 0),
                                                                    "bulk\neTAC"=c(0, 0.5)))
# png("~/Dropbox/ETACS/plots.dir/bulk_eTAC_vs_mTEC-TRAs_Venn.png",
#     height=8.5, width=8.5, res=300, units="in")
grid.draw(thway.venn)
# dev.off()
```

I've also compared the TRAs detected in the cDC, mixed bulk eTAC and mTEC samples.  There are not any TRAs specific to just the bulk eTACs or the cDCs, but there are 18 that are shared between them.  This adds weight to the conclusion that these cells do not actively express endogenous TRAs as hypothesised.

```{r, echo=FALSE}
not.cDC <- setdiff(intersect(mtec.tras, bulk_etac.tras), cdc.tras)
gene_symbol <- getBM(attributes=c('ensembl_gene_id', 'external_gene_name'),
                     filters='ensembl_gene_id', mart=ensembl,
                     values=not.cDC)
write.table(gene_symbol, file="~/Dropbox/Reports_eTAC/bulk-eTAC_intersect_mTEC.tsv",
            sep="\t", row.names=F, quote=F)
gene_symbol
```

Are there any functional enrichments within the 167 genes shared between eTACs and mTECs, but not cDCs?

```{r, echo=FALSE, fig.height=5.5, fig.width=12.5, warning=FALSE, message=FALSE}
all.genes <- etacs$gene_id
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
gene.tpm <- read.table("~/Dropbox/ETACS/human_etacs/eTAC-8-1.quant", sep="\t",
                       h=T, stringsAsFactors=F)
gene.length <- gene.tpm$Length[unique(gene.tpm$Name) %in% unique(all.genes)]
names(gene.length) <- intersect(unique(gene.tpm$Name) , unique(all.genes))

de.big.vector <- as.integer(unique(all.genes) %in% unique(not.cDC))
names(de.big.vector) <- unique(all.genes)

de.big.vector <- de.big.vector[intersect(names(de.big.vector), names(gene.length))]

pwf.all <- nullp(de.big.vector, "hg38", "ensGene", bias.data=gene.length[intersect(names(de.big.vector), names(gene.length))],
                 plot.fit=FALSE)
GO.wall.all <- goseq(pwf.all, genome="hg38", id="ensGene", test.cats=c("GO:BP"),
                      method="Wallenius", use_genes_without_cat=T)

GO.wall.all$padjust <- p.adjust(GO.wall.all$over_represented_pvalue, method="BH")
GO.wall.all$foldEnrich <- ((GO.wall.all$numDEInCat/length(unique(not.cDC)))/(GO.wall.all$numInCat/length(gene.length)))

# select on p-value and foldEnrich
GO.sig.all <- GO.wall.all[(GO.wall.all$padjust <= 0.01), ]
go_cats.all <- data.frame(GO.sig.all$category, GO.sig.all$foldEnrich)

p.enrich <- ggplot(GO.sig.all,
                   aes(x=foldEnrich, y=-log10(padjust), colour=-log10(padjust),
                       size=foldEnrich)) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C")

GO.all <- GO.sig.all[order(GO.sig.all$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(GO.all[1:20, ]),
                  aes(x=reorder(term, 
                                foldEnrich),
           y=foldEnrich,
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="Gene Ontology Term", y="Fold Enrichment")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

Nope, no functional enrichments.  Are they enriched for particular tissues?  We might expect to see enrichment of non-immune tissues amongst shared mTEC and eTAC genes if there is genuine pGE in the bulk eTAC cells.  Conversely, we should __not__ see any tissue enrichments in the genes specific to the conventional DCs.

I will look at the intersections of all genes, not just the TRAs, as we want to know if there is a general enrichment compared to the whole transcriptome, not just to other TRAs.  This is because the definition tissue-restricted will be biased by the coverage of different cell types and tissues in the FANTOM5 data.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
tra.df <- read.table("~/Dropbox/fantom5/hg19/Human-TRA-FANTOM5-Tissues.tsv",
                     h=T, stringsAsFactors=F, sep="\t")
```

```{r, echo=FALSE}
cdc.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("cDC")])
bulk_etac.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("bulk_eTAC")])
bulk_mtec.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("human_mTEC")])
```


```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
tissues <- unique(tra.df$Tissue)
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
# the background set of genes needs to be all TRAs?
top <- setdiff(intersect(bulk_etac.tras, mtec.tras), cdc.tras)
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)
top.bulk_etacs <- top

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

This analysis suggests that there is some specific tissue enrichment amongst the `r length(top)` TRAs shared between bulk mTECs and eTACs, but not cDCs.  Given the heterogeneity in these bulk samples I am concerned this might just be an artifact.  To demonstrate the validity of this approach, the plots below show the enrichment of many more tissues amongst the genes unique to mTECs.


```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(mtec.tras, union(cdc.tras, bulk_etac.tras))
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)


for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

As a negative control, we should see an absence of enrichments amongst the cDC-specific genes with respect to non-immune tissues.

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(cdc.tras, union(mtec.tras, bulk_etac.tras))
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)


for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

Good, there are no tissue-enrichments in the cDC-specific TRAs.

The presence of a few tissue enrichments does perhaps indicate there is some biologically interesting TRA expression in the bulk eTACs, though this is caveated by the fact this is known to be a heterogenous mix of cell states which may confound this analysis.  I'll repeat this for the single cell groups from the pseudotime analysis to see if a particular group is likely responsible for the observed enrichments.  Given the current evidence, I suspect this would be the `yellow` group of eTACs, which look like monocyte/macrophages.  Ideally we would like to see an enrichment in the `brown` eTAC group, as these are the ones that express _AIRE_.

```{r, echo=FALSE}
yellow_etac.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_yellow")])
blue_etac.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_blue")])
brown_etac.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_brown")])
turq_etac.genes <- unique(fantom.merge$gene_id[fantom.merge$value > 1 & fantom.merge$CellType %in% c("eTAC_turquoise")])
```


### Blue cells tissue enrichment testing

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(intersect(blue_etac.tras, mtec.tras), unique(c(yellow_etac.tras, brown_etac.tras, turq_etac.tras)))
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

There is only a depletion of non-TRAs (hardly surprising as these are all TRAs).

### Turquoise cells tissue enrichment testing

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(intersect(turq_etac.tras, mtec.tras), unique(c(yellow_etac.tras, brown_etac.tras, blue_etac.tras)))
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

Well, there's a depletion for non-TRA genes, but no specific enrichment for any tissues...

### Yellow cells tissue enrichment testing

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(intersect(yellow_etac.genes, mtec.tras), unique(c(turq_etac.genes, brown_etac.tras, blue_etac.tras)))
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```


Again, only a depletion of non-TRA genes.

### Brown cells tissue enrichment testing

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(intersect(brown_etac.genes, mtec.tras), unique(c(turq_etac.genes, yellow_etac.genes, blue_etac.tras)))
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

There isn't even a depletion of TRAs in the `brown` cells.  No evidence for any strong enrichments for any of the single eTAC groups for particular tissue groups.  We've seen that the mTECs are particularly enriched for lots of different tissues, but what about the combined single eTACs?  Do they create spurious enrichments when combined together?


### All single eTAC cells tissue enrichment testing

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(intersect(mtec.tras, c(brown_etac.tras, yellow_etac.tras, turq_etac.tras, blue_etac.tras)), cdc.tras)
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.05, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```
This might indicate the possible false positive result that could be obtained by combining heterogenous cells together in this sort of enrichment analysis.  How much should one read into the consistency for the enrichment of lung TRAs between the bulk eTACs and combined single eTAC groups?  The Lung category is the 5th largest after `Testis`, `Macrophage`, `Intestine` and `Hepatocyte`.

What are the TRAs that are the same between bulk eTACs and combined single eTACs, and are they still enriched for `Lung` TRAs.

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top.bulk_etacs <- setdiff(intersect(bulk_etac.tras, mtec.tras), cdc.tras)
top_single_etacs <- setdiff(intersect(mtec.tras, c(brown_etac.tras, yellow_etac.tras, turq_etac.tras, blue_etac.tras)), cdc.tras)
 top <- intersect(top_single_etacs, top.bulk_etacs)
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.05, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

So the TRAs in both the bulk eTACs and single eTACs are not even the `Lung` TRAs.  This really makes me think that these enrichments arise purely by chance/as a consequence of combining heterogenous populations.


```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
#####################################################################
## Tissue enrichment testing for eTACs and thymus single cell data ##
#####################################################################
fisher.list <- list()
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- setdiff(union(cdc.tras, bulk_etac.tras), mtec.tras)
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.05, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

Monocytes and macrophages probably close enough given the mTECs express pretty much everything anyway, and I don't think that there are any specific DCs in the FANTOM5 data.

As a final sanity check, it would be useful to see what the tissue enrichments are for the pancreatic $\alpha$ islets single cells is.

### Pancreatic $\alpha$ islets tissue enrichment testing

```{r, echo=FALSE, fig.width=9.5, fig.height=5.5}
fisher.list <- list()
panc.genes <- panc.alpha[rowMeans(panc.alpha[, 1:(dim(panc.alpha)[2]-1)]) > 1, ]$gene_id
# all.genes <- unique(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"])
top <- intersect(fantom.merge$gene_id[fantom.merge$GeneGroup == "Tissue-Restricted"], panc.genes)
all.genes <- setdiff(unique(fantom.merge$gene_id[fantom.merge$GeneGroup != "Constituitive"]), top)

for(i in 1:length(tissues)){
  tiss <- tissues[i]
  side <- tra.df$ensembl.gene[tra.df$Tissue == tiss]
  
  a <- length(intersect(top, side))
  b <- length(intersect(all.genes, side))
  c <- length(setdiff(top, side))
  d <- length(setdiff(all.genes, side))
  
  if(a == 0){
    a <- 1
  }
  
  if(b == 0){
    b <- 1
  }
  
  if(c == 0){
    c <- 1
  }
  
  if(d == 0){
    d <- 1
  }
  
  fish.res <- fisher.test(matrix(c(a, c, b, d), ncol=2), or=1)
  P <- fish.res$p.value
  OR <- fish.res$estimate
  
  SE <- sqrt(sum(c(1/a, 1/b, 1/c, 1/d)))
  fisher.list[[tiss]] <- c("OR"=OR, "pvalue"=P, "SE"=SE)
}

fisher.df <- data.frame(do.call(rbind, fisher.list))
colnames(fisher.df) <- c("OR", "pvalue", "SE")
fisher.df$padjust <- p.adjust(fisher.df$pvalue)
fisher.df$Tissue <- rownames(fisher.df)

fisher.sig <- fisher.df[fisher.df$padjust <= 0.01, ]

p.enrich <- ggplot(fisher.sig,
                   aes(x=log(OR), y=-log10(padjust), colour=-log10(padjust),
                       size=log(OR))) +
  geom_point() + theme_mike() +
  scale_colour_gradient(low="#000000", high="#DC143C") +
  labs(x="log Odds Ratio", y="-log10 p-value")

fisher.all <- fisher.sig[order(fisher.sig$padjust, decreasing=FALSE), ]

p.terms <- ggplot(na.omit(fisher.all),
                  aes(x=reorder(Tissue, 
                                OR),
           y=log(OR),
           fill=-log10(padjust))) +
  scale_fill_distiller(palette="Reds", direction=1) +
  geom_bar(stat='identity') + theme_mike() +
  coord_flip() + labs(x="FANTOM5 Tissue", y="log Odds Ratio")

plot_grid(p.enrich, p.terms, ncol=2, rel_widths=c(0.8, 1.3))
```

Excellent, there is at least an enrichment amongst the pancreas TRAs in several endocrine organs and others with similar ontogeny as the pancreas.  I'm not sure about the enrichment for 3 different brain regions though, this might be due to how the genes for these cells were selected.
